### PR TITLE
EES-XXXX - amending run_tests.py to detect failed suites from output.xml

### PR DESCRIPTION
**Pull request by duncan-at-hiveit**
image

During some battles with text selector changes, I broke the Robot code a few times but noticed that misconfigured keywords and errors within our own keyword code didn't cause our tests to fail fast (nothing new, as the `run_on_failure` that we use to trigger failures only kicks in within SeleniumLibrary's own keywords as has always been the case). This would mean that some suites ultimately failed but didn't get recorded by our fail-fast mechanism.

They were however captured in output.xml and so I changed the code up to use this as the ultimate source of truth instead, should we accidentally introduce keyword changes that break them but don't get caught by our fail-fast mechanism.

Ultimately to fix the fail-fast issue we could do with setting `run_on_failure=None` to prevent Selenium-only keywords failing suites fast, and instead in each test suite specify `Test Teardown    Run Keyword On Test Failure    Record failing suite` - the only issue here being that it doesn't catch failures in `Suite Setup`, so we'd need another mechanism for that, potentially with listeners.

## Additional console logging details

There is now a direct link to the last run attempt's report `log.html` as well as the merged report, meaning you can jump to the final run's failures without all the noise of the full test suite results.

```
2024-10-10 14:20:40,753: *************************************
2024-10-10 14:20:40,753: FINAL REPORT
2024-10-10 14:20:40,753: Log available at: file:///home/dwatson/development/projects/dfe/explore-education-statistics/tests/robot-tests/test-results/log.html
2024-10-10 14:20:40,753: Report available at: file:///home/dwatson/development/projects/dfe/explore-education-statistics/tests/robot-tests/test-results/report.html
2024-10-10 14:20:40,753: *************************************

2024-10-10 14:20:40,753: *************************************
2024-10-10 14:20:40,753: LAST RUN ATTEMPT REPORT
2024-10-10 14:20:40,753: Log available at: file:///home/dwatson/development/projects/dfe/explore-education-statistics/tests/robot-tests/test-results/run-1/log.html
2024-10-10 14:20:40,753: Report available at: file:///home/dwatson/development/projects/dfe/explore-education-statistics/tests/robot-tests/test-results/run-1/report.html
2024-10-10 14:20:40,753: *************************************

2024-10-10 14:20:40,753: Number of test run attempts: 1
2024-10-10 14:20:40,753: 
All tests passed!
```

### CODE REVIEW

File: tests/robot-tests/reports.py [hunk 1]
```diff
@@ -79,36 +105,48 @@ def _merge_test_reports(test_results_folder):
...
+    suite_results = report.find("statistics").find("suite").find_all("stat")
+    failing_suite_results = [suite_result for suite_result in suite_results if int(suite_result.get("fail")) > 0]
+    suite_ids = [failing_suite.get("id") for failing_suite in failing_suite_results]
+    leaf_suite_ids = []
+    for suite_id in suite_ids:
+        other_suite_ids_containing_suite_id = [
+            other_suite_id for other_suite_id in suite_ids if other_suite_id != suite_id and suite_id in other_suite_id
+        ]
+        if len(other_suite_ids_containing_suite_id) == 0:
+            leaf_suite_ids += [suite_id]
```
Comment (mmyoungman) [inline tests/robot-tests/reports.py:143]: Is this to remove duplicates from `suite_ids`? I think you could do `suite_ids_no_dups = list(set(suite_ids))`

File: tests/robot-tests/reports.py [hunk 2]
```diff
@@ -79,36 +105,48 @@ def _merge_test_reports(test_results_folder):
...
-    passing_suite_results = [suite_result for suite_result in suite_results if int(suite_result.get('fail')) == 0]
-    return [passing_suite.get('id') for passing_suite in passing_suite_results]
+
+def _get_passing_suite_ids_from_report(report: BeautifulSoup) -> []:
+    suite_results = report.find("statistics").find("suite").find_all("stat")
+    passing_suite_results = [suite_result for suite_result in suite_results if int(suite_result.get("fail")) == 0]
+    return [passing_suite.get("id") for passing_suite in passing_suite_results]
+
+
+def _get_failing_leaf_suite_ids_from_report(report: BeautifulSoup) -> []:
```
Comment (mmyoungman) [inline tests/robot-tests/reports.py:133]: I'm not sure what you mean by `leaf` here

**Review by mmyoungman — APPROVED** [review]

File: tests/robot-tests/tests/libs/fail_fast.py [hunk 3]
```diff
@@ -93,8 +94,8 @@ def _prompt_to_continue():
     logger.warn("Continue? (Y/n)")
     choice = input()
     if choice.lower().startswith("n"):
-        raise_assertion_error("Tests stopped!")
-        
+        _raise_assertion_error("Tests stopped!")
+
 
 def _raise_assertion_error(err_msg):
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply tests/robot-tests/tests/libs/fail_fast.py:100]: Oh no, it still is:

```
def _raise_assertion_error(err_msg):
    sl().failure_occurred()
    raise AssertionError(err_msg)
```

This is different code than the skipping of tests :)

File: tests/robot-tests/reports.py (hunk 1 above)
Comment (duncan-at-hiveit) [reply tests/robot-tests/reports.py:143]: This is to remove non-leaf suites.  The error report xml includes stats reporting at multiple levels, so if you have:

Admin:Admin And Public:Some Leaf Suite Here

Comment (duncan-at-hiveit) [reply tests/robot-tests/reports.py:143]: Given it wasn't very obvious, I've added a comment to this effect in the code.

File: tests/robot-tests/reports.py (hunk 2 above)
Comment (duncan-at-hiveit) [reply tests/robot-tests/reports.py:133]: See above :)

File: tests/robot-tests/tests/libs/fail_fast.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmyoungman) [reply tests/robot-tests/tests/libs/fail_fast.py:100]: My bad - I didn't see the 3rd line...

**Pull request merged**